devcontainer/devcontainer.go

```diff
@@ -66,8 +66,50 @@ type Compiled struct {
 	BuildContext      string
 	BuildArgs         []string
 
-	User string
-	Env  []string
+	User         string
+	ContainerEnv map[string]string
+	RemoteEnv    map[string]string
+}
+
+func SubstituteVars(s string, workspaceFolder string) string {
+	var buf string
+	for {
+		beforeOpen, afterOpen, ok := strings.Cut(s, "${")
+		if !ok {
+			return buf + s
+		}
+		varExpr, afterClose, ok := strings.Cut(afterOpen, "}")
+		if !ok {
+			return buf + s
+		}
+
+		buf += beforeOpen + substitute(varExpr, workspaceFolder)
+		s = afterClose
+	}
+}
+
+func substitute(varExpr string, workspaceFolder string) string {
+	parts := strings.Split(varExpr, ":")
+	if len(parts) == 1 {
+		switch varExpr {
+		case "localWorkspaceFolder", "containerWorkspaceFolder":
+			return workspaceFolder
+		case "localWorkspaceFolderBasename", "containerWorkspaceFolderBasename":
+			return filepath.Base(workspaceFolder)
+		default:
+			return os.Getenv(varExpr)
+		}
```

Comment:
It'd be helpful to link the reference document here in case future var expressions are added!